### ISSUE TITLE
[BUG] Issues Scaling Down RKE2 Cluster w/ Rancher v2.9.3-alpha5 (upgraded from v2.9.2) & Harvester v1.3.2 w/ Harvester Node Driver v0.7.0

### ISSUE THREAD

**Issue by irishgordo** [harvester/harvester#6775]
#### Describe The Bug
Using Node Driver v0.7.0 - there are some underlying issues trying to scale down a 3 node (all nodes etcd/cp/worker) to 1 -> after upgrading Rancher from v2.9.2 -> v2.9.3-alpha5.

#### To Reproduce:
- have either a K3s or RKE2 cluster w/ Rancher v2.9.2 installed
- note: this was tested w/ a Harvester VM running Jammy Cloud-Img, RKE2 v1.3.0.5+rke2r1, quick installed
1. download locally [v0.7.0](https://github.com/harvester/docker-machine-driver-harvester/releases/tag/v0.7.0) Harvester Node Driver aka docker-machine-driver-harvester , `tar xvf` that tarball, to get the `docker-machine-driver-harvester` binary
2. `kubectl get pods -n cattle-system`, finding your v2.9.2 Rancher pod
3. `kubectl cp docker-machine-driver-harvester $rancher-pod-name:/usr/share/rancher/ui/assets/docker-machine-driver-harvester`
4. validate sha256sums on your docker-machine-driver-harvester versus the one available in assets so just https://rancher.192.168.104.204.sslip.io/assets/docker-machine-driver-harvester , replace ipv4 with the correct one for your Rancher v2.9.2 instance -> ensure the docker-machine-driver-harvester `sha256sum` matches that of the v0.7.0 
5. import Harvester into Rancher
6. set in Rancher `kubeconfig-default-token-ttl` to `60` minutes
7. spin up a 3 node rke2 cluster
8. `helm repo update` , make sure you have the `rancher-alpha/rancher` repo available for helm
9. `helm get values rancher -n cattle-system -o yaml > values.yaml`
10. `helm upgrade rancher rancher-alpha/rancher --devel --namespace cattle-system -f values.yaml --version=2.9.3-alpha5`
11. `kubectl -n cattle-system rollout status deploy/rancher`
12. wait for the rollout to be complete
13. again, find the Rancher pod `kubectl get pods -n cattle-system | grep -ie "rancher"`
14. again, copy over the v0.7.0 Harvester Node driver to the Rancher pod `kubectl cp docker-machine-driver-harvester $rancher-pod-name:/usr/share/rancher/ui/assets/docker-machine-driver-harvester`
15. try to scale down the RKE2 cluster in Rancher v2.9.3-alpha5 from 3 nodes to 1 node
16. there seems to be an issue, the cluster is hanging

**Expected behavior**
scaling down to work appropriately

**Environment**
 - Harvester ISO version: v1.3.2
 - Underlying Infrastructure: Bare-Metal, HP DL160 Gen9
 - Rancher: v2.9.2 & v2.9.3-alpha5
 - RKE2 running Rancher: v1.30.5+rke2r1
 - VM Running RKE2 Running Rancher: 50Gi Disk virtio, 4C, 8Gi (On Harvester), jammy-server-cloudimg-amd64-disk-kvm.img

**Comment by irishgordo** [comment]
This might have some relation to: https://github.com/harvester/harvester/issues/4358

**Comment by m-ildefons** [comment]
You may want to try again with Rancher v2.9.3-alpha6.
The alpha5 version didn't yet contain the Harvester node driver v0.7.0, so I suspect that your upgrade of Rancher to the v2.9.3-alpha5 version accidentally downgraded the Harvester node driver again.
~Alternatively, you may try again, but do the `kubectl cp` step again after upgrading Rancher.~

Edit: I just saw you already tried to `kubectl cp` again... 
But I do wonder if your reproducer relies on the Kubeconfig token expiring, or if that's unrelated. Setting the TTL to 60 minutes shouldn't do anything unless you also wait for the TTL to pass.

You may also have to replace the binary at
`/opt/drivers/management-state/bin/docker-machine-driver-harvester` with the v0.7.0 binary

**Comment by m-ildefons** [comment]
If you read the Dockerfile of Rancher, `/usr/share/rancher/ui/assets/docker-machine-driver-harvester` should just be a symlink to `/opt/drivers/management-state/bin/docker-machine-driver-harvester`. But there is also another copy of the node driver at `/opt/jail/driver-jail/var/lib/rancher/management-state/bin/docker-machine-driver-harvester`. No idea why that is though...

There is also the `nodedriver.management.cattle.io/v3` CRD, which may need to be modified to point to the new driver version

**Comment by m-ildefons** [comment]
I just ran two simplified versions of the scenario in question:

1) Isolated test of the scale-down functionality in the Harvester node driver v0.7.0

- Set up Harvester (v1.3.1) and Rancher (v2.9.2)
- Replace Harvester node driver binaries in the Rancher container (in all three locations)
- Edit the `nodedriver.management.cattle.io/v3` resource
- Create 3 node RKE2 guest cluster
- Scale down guest cluster to 1 node

2) Isolated test of cloud credential rotation

- In the same Harvester/Rancher setup as above, still with the manually replaced Harvester node driver
- Create new cloud credential with 1h expiry time
- Create new 3 node RKE2 guest cluster
- Wait for cloud credential to expire
- Rotate credentials according to KB article instructions
- Scale down RKE2 guest cluster to 1 node

Both scenarios successfully completed and the RKE2 guest clusters ended up as healthy 1 node clusters. This leads to the conclusion that the node driver v0.7.0 works ok and the reported problem is a result of improper/incomplete replacement of the node driver.

It should be noted that in the second scenario, the scaling of the guest cluster took a considerable time and the guest cluster seemed to be stuck half-deleted for about 30 minutes before the scaling-down finally proceeded.
While not ideal, I don't consider this a bug in the node driver per se.

In summary, when manually replacing the node driver, make sure:
- Replace symlink at `/usr/share/rancher/ui/assets/docker-machine-driver-harvester`
  This symlink should point to `/opt/drivers/management-state/bin/docker-machine-driver-harvester` or be replaced with a copy of the node driver binary
- Replace binary at `/opt/drivers/management-state/bin/docker-machine-driver-harvester`
- Replace binary at `/opt/jail/driver-jail/var/lib/rancher/management-state/bin/docker-machine-driver-harvester`
- Patch `nodedriver.management.cattle.io/v3` CRD, specifically the `.spec.url`, `.spec.checksum` and `.spec.whitelistDomains` fields

**Comment by m-ildefons** [comment]
I also ran a third simplified test scenario using Harvester v1.3.2 and Rancher v2.9.3-alpha6, just testing if credential rotation works there as expected:

- Set up Harvester v1.3.2 and Rancher v2.9.3-alpha6
- Configure short token TTL (1h)
- Create 3 node RKE2 guest cluster
- Wait for token to expire
- Apply workaround according to KB article
- Scale guest cluster down to 1 node

This also worked successfully. I'm going to hold off on testing further scenarios, which include upgrades until https://github.com/rancher/rancher/issues/47583 has been fixed.

**Comment by irishgordo** [comment]
Closing this out as this was successful on the scenarios when the node driver wasn't needing to be manually adjusted ( I believe the original issue there was an issue in the configuration of manually patching the node-driver ) - as most users will not be trying to manually patch the node driver.  When not needing to manually patch in testing this looked good :smile: .   And since 47583 (and some others relating to that) have been merged. 
Feel free to re-open if there's an edge scenario that wasn't accounted for - thanks for helping also check this @m-ildefons :smile: :+1:
